jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/FeatureWriter.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.jbellis.jvector.graph.disk;
+
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * A feature of an on-disk graph index. Information to use a feature is stored in the header on-disk.
+ * This interface extends {@link Feature} to add the ability to write the feature inline, which
+ * may require more context than is stored on-disk.
+ */
+interface FeatureWriter extends Feature {
```

Comment:
it looks like the distinction is we only need Feature at read time, while the Writers need additional state

so it probably makes more sense to push writeHeader into FeatureWriter instead of trying to eliminate FeatureWriter

Reply:
Looks like the only code path that calls writeHeader has a FeatureWriter already, so I went ahead and did this in my branch (with a cast).